src-java/kilda-pce/src/main/java/org/openkilda/pce/finder/BestWeightAndShortestPathFinder.java

```diff
@@ -448,33 +449,72 @@ private SearchNode getDesiredPath(Node start, Node end, WeightFunction weightFun
             }
 
             // Shift the current weight relative to maxWeight
-            long shiftedCurrentWeight = Math.abs(maxWeight - current.parentWeight.toLong());
+            long shiftedCurrentWeight = Math.abs(maxWeight - current.parentWeight.getBaseWeight());
 
             // Determine if this node is the destination node.
             if (current.dstSw.equals(end)) {
                 // We found the destination
-                if (shiftedCurrentWeight < desiredWeight && current.parentWeight.toLong() < maxWeight) {
+                if (current.parentWeight.getBaseWeight() < maxWeight) {
                     // We found a best path. If we don't get here, then the entire graph will be
                     // searched until we run out of nodes or the depth is reached.
-                    desiredWeight = shiftedCurrentWeight;
-                    desiredPath = current;
+                    if (currentPathWeight.getPenaltyValue(PathWeight.Penalty.DIVERSITY_ISL_LATENCY) > 0
+                            || currentPathWeight.getPenaltyValue(PathWeight.Penalty.DIVERSITY_SWITCH_LATENCY) > 0
+                            || currentPathWeight.getPenaltyValue(PathWeight.Penalty.DIVERSITY_POP_ISL_COST) > 0) {
```

Comment:
I'm not sure about this `if`

Could you please add tests with this penalties and check that results are same with previous behavior?

Reply:
Sure! Tests were added and logic changed a bit.